test/types/range/unbounded/unboundedIter-3.good

```diff
@@ -0,0 +1,4 @@
+121
+123
+125
+unboundedIter.chpl:13: error: halt reached - Loop over unbounded range surpassed representable values
```

Comment:
It seem a little surprising / unfortunate that this range won't yield `-127` when `int(8)` includes `-128`.  `-127` seems representable, and is represented in the `-4.good` output.

Reply:
I'm not understanding... you're saying that you think iterating over `121:int(8).. by 2` should yield -127?

As noted in the OP, I also wanted to yield all values represented by the idxType—and had drafts of the PR that did so—but could not come up with a way of doing it that didn't require an extra yield, which disabled our zippered iteration optimizations.

Reply:
Err, oops.  I meant to attach that to `-6.good`, for `..-121:int(8) by -2 `, which I'd have expected to include the -127.  Sigh.

Reply:
The essential challenge is that we generate a C for loop along the lines of:

```c
for (i=start; i >= stop; i+= stride) { yield i; }
```

so for this loop:

```c
int8_t i;
for (i=-121; i >= stop; i+= -2) { yield i; }
```

so if we used `-127` for `stop`, our next += -2 would wrap us around to a positive value and we'd get the infinite loop that we've traditionally had (since that value is >= -127).  The way I tried to deal with this when developing this branch was to add an extra yield outside of the loop that would take care of this boundary value.  So, something like:

```c
if (i == -127) {
  yield i;
}
```

and this worked in terms of giving the correct/preferable results, but it had the downside of breaking our loop zippering optimization for cases like `for (i,j) in zip(whatever, 0..)` because our compiler currently doesn't know how to efficiently zip an arbitrary single-loop iterator (the common case for 1D loops) with a 1D loop that has one additional yield outside of the loop.  Because we rely on this idiom in some key codes, it seemed better to stop the iteration one value early (esp. given that break-free loops over unbounded ranges are relatively rare) than to hurt performance for all such zippered loops.

I think the right way to address this going forward is to have distinct serial iterators for the standalone vs. zippered context, like we do for parallel iterators.  This would permit us to keep the zippered case efficient (in fact, it could even arguably use the "infinite" implementation if we did a size check to make sure it was legal since it will be following a likely finite iterator) while having the standalone case use this more precise "extra yield" approach (since it doesn't hurt performance in the non-zippered context.

Another way we could've dealt with this would be to use a larger int type for the `i` in loops like this and to downcast on the yield, but I wasn't willing to do that here.  Primarily because it seems philosophically objectionable to me to have a computation that strives hard to use small ints and to have larger ints get used under the covers (if, say, someone was trying to run this on an embedded platform or the like and had bit width constraints).

Finally, I take some small comfort in the fact that since the iterator would've halted right after yielding the -127 (i.e., useful programs couldn't do it anyway), nothing significant seems lost by halting one step early.  Combined with the unlikelihood of iterating to the extremes of an unbounded iterator to begin with for any type other than int(8)/uint(8) (?).  Though I'll be curious if users find otherwise and complain.

Reply:
> the unlikelihood of iterating to the extremes of an unbounded iterator to begin with for any type other than int(8)/uint(8) (?).

Any sized uint would hit this for `..10:uint by -1`, etc.

(Kinda related, similarly for bounded ranges `0..10:uint by -1`.  It further seems unsatisfying and asymmetric that you can iterate over `0..10:uint by 1`, but not by `-1`.)

> I think the right way to address this going forward ...

Are there any concerns with changing the behavior as suggested above after a release?   I guess since all the bad behavior causes halts (right?), no program will be depending on it.


One way that comes to mind would be to calculate the exact value that will be hit one step after `stop`.  For stride `1` or `-1`, that's easy, `stop+1` or `stop-1`, so

```
for (i=start; i != reallyStop; i+= stride) { yield i; }
```
Even if `reallyStop` has over/underflown the type, we'll still hit it.  (As long as `start != reallyStop`, which I think would imply a maximal range that's already not supported?)

For non-unit strides, it would take some more arithmetic to find the end point. And it could run into `start == reallyStop` for non-maximal ranges like `0..128:uint(8) by 128` or `0..192:uint(8) by 64`.


Another way that might have too much overhead would be to find a safe shift value and do

```
for (i=start+shift; i<=stop+shift; i+= stride) { yield i-shift; }
```

Or figure out how many iterations there will be, and

```
for (i=start, n=0; n<count; i+= stride, n++) { yield i; }
```
(These last two both require two extra operations per iteration.)

Reply:
> Any sized uint would hit this for ..10:uint by -1, etc.

True, but my point is more like "How often do people want to iterate over unbounded ranges with the intent of falling off the end of the legally representable values?  I.e., why would I write the above set of integers as "negative infinity to 10 by -1" rather than `0..10:uint by -1`? (oh, though you point out we can't actually write that today...  though with ints we could, and the index could be cast after the fact).  Particularly since, even if we did yield the 0, the next thing we'd do is halt your program for falling off the end?  Or were you going to put `if i == 1 then break;` into your loop to avoid that?  (in which case, why not write `1..10:uint by -1` instead?)  Am I making sense?

> (Kinda related, similarly for bounded ranges 0..10:uint by -1. It further seems unsatisfying and asymmetric that you can iterate over 0..10:uint by 1, but not by -1.)

I agree, and we have some open issues about "maximal ranges" (or, better, ranges that rely on iterating up to maximal values) that serve as placeholders for not having a completely satisfying story here.

> I guess since all the bad behavior causes halts (right?), no program will be depending on it.

Right, that's my thinking.  And since, previously, those loops were infinite, I think we're in a strictly better place than we used to be.

> Even if reallyStop has over/underflown the type, we'll still hit it.

Oh, that's an interesting idea.  Will that work in all cases, though?  E.g., aren't wraparounds only well-defined on unsigned types in C and not signed integers?  Still, that could be better than nothing.  I'm trying to remember whether we had an implementation that did this before but wasn't working for some other reason or resulted in worse performance.  That said, I can't think of a reason that it wouldn't.  Tagging @ronawho on this who did some early work on ranges with these maximal values to see what he remembers.

> (These last two both require two extra operations per iteration.)

Yeah, I feel less inclined to take those approaches for fear of making all loops a bit more heavyweight to cater to some cases that seem less common to me.  We've talked about having some compiler flags that would opt into supporting tricky cases by hook or by crook, and could consider generating these more complex forms when such a flag was used, e.g.  Mostly, though, nobody in the user community has seemed bothered by the current gaps (that I'm aware of anyway).  Even the infinite loop over unbounded ranges is something that I'm not aware of any users noticing or complaining about.

Reply:
> True, but my point is more like "How often do people want to iterate over unbounded ranges with the intent of falling off the end of the legally representable values? I.e., why would I write the above set of integers as "negative infinity to 10 by -1" rather than `0..10:uint by -1`? (oh, though you point out we can't actually write that today... though with ints we could, and the index could be cast after the fact). Particularly since, even if we did yield the 0, the next thing we'd do is halt your program for falling off the end? Or were you going to put `if i == 1 then break;` into your loop to avoid that? (in which case, why not write `1..10:uint by -1` instead?)

Really, I was just thinking it was surprising that it was halting before yielding the last representable value (`-127` for `int(8)`, or `0` here.)

I could maybe contrive a case where we're searching for an index that has some property, and we know some index with `abs(i) < 10` will satisfy it, but the function is generic over the index type so we can't write either `0..10` or `-10..10`.  

So rather than write out the cases by type, I would have expected this to work, even when 0 is the unsigned value we're looking for:

```chapel
for i in ..endPoint by -1 {
  if (found_something(i)) {
     do_it(i);
     break;
  }
}
```

I was imagining that if the uint iterator could yield 0, then the loop body would have a chance to break out of the loop before the next entry to the iterator halted due to overflow.  So that it wouldn't be the intention of falling off the end, but iterating over the legally representable values.


Maybe this pattern would be better off with a C-style loop rather than iterating over a range.  Or using the `generalIterator()`, which I'd been unaware of.  Maybe the range primer should mention it and suggest when you should use it?

(Heh, the `generalIterator()` appears to implement the same two-variable algorithm @ghbrown proposed; and has a comment contemplating the "figure out how many iterations there will be" approach I wrote above.)

Reply:
> I could maybe contrive a case where we're searching for an index that has some property

This is exactly what I was trying to do as part of solving [Advent of Code 2015 Day 4](https://adventofcode.com/2015/day/4), which led me to this discussion. In fairness, I never intended to fall off the end (I don't think anybody really does, but they may certainly be interested in the endpoint value).